Add support for batching pachctl PFS commands into transactions

Tryneus opened the pull request:
This PR adds several new pachctl commands:
 * `pachctl start transaction`
 * `pachctl stop transaction`
 * `pachctl resume transaction`
 * `pachctl finish transaction`
 * `pachctl list transaction`
 * `pachctl inspect transaction`
 * `pachctl delete transaction`

Transactions are stored in etcd with a list of requests and responses.  Any time a new request is added to the transaction, the entire transaction is dry-run on the current state of the cluster metadata to make sure it is valid.  If successful, the transaction will be updated, otherwise the first error is returned to the client.  When the transaction is `pachctl finish`ed, the transaction is run to completion and commits the new etcd state, and the transaction itself is removed from the cluster.

On the client side, the currently active transaction ID is stored in the user's local pachyderm config file.  The following commands will check if there is an active transaction and append to it rather than issuing the command directly:

* `pachctl create repo`
* `pachctl update repo`
* `pachctl delete repo`
* `pachctl start commit`
* `pachctl finish commit`
* `pachctl delete commit`
* `pachctl create branch`
* `pachctl delete branch`
* `pachctl copy file`
* `pachctl delete file`

In follow-up PRs, I intend to extend this behavior to more commands.  They were not included here to limit the scope of this PR and due to extra complications with these commands:
* `pachctl put file`
  * This needs to be done in two stages - uploading the file before the transaction and updating the metadata within the transaction.
  * Also some considerations to prevent garbage collecting uploaded files while the transaction is open.
* PPS commands:
  * These need to be revamped to perform PFS commands transactionally using the newly exposed PFS APIServer interface, then PPS can support transactions itself.
  * `pachctl create pipeline`
  * `pachctl update pipeline`
  * `pachctl delete pipeline`
  * `pachctl start pipeline`
  * `pachctl stop pipeline`
* `pachctl delete all`
* possibly auth stuff

Still TODO before this PR is ready for review:
* [x] Finish implementing the pretty-printing of transaction details (print individual requests in the transaction)
* [x] Add more tests

Tryneus commented on src/server/pfs/server/driver.go:
```diff
@@ -1340,278 +1374,281 @@ func (d *driver) deleteCommit(pachClient *client.APIClient, userCommit *pfs.Comm
 	// TODO put scratch spaces in a collection and do this in the txn above
+	// TODO: probably have to do that before merging this
+	// TODO: do this in a defer because otherwise this probably leaks, also probably
+	// not safe to do this inside of an stm transaction
```
Looks like I still need to take care of this TODO.

Tryneus commented on src/server/pfs/server/driver.go:
```diff
@@ -3036,21 +3070,22 @@ func (d *driver) deleteFile(pachClient *client.APIClient, file *pfs.File) error
+		_, err := d.makeCommit(pachClient, stm, "", client.NewCommit(file.Commit.Repo.Name, ""), branch, nil, nil, []string{file.Path}, []*pfs.PutFileRecords{&pfs.PutFileRecords{Tombstone: true}}, "")
 		return err
 	}
+	// TODO: what do with this in STM transaction?
```
Also this.

Tryneus replied:
I'm thinking I can just drop this TODO - `upsertPutFileRecords` doesn't appear to ever be done in a transaction, but I'm curious if this could cause any problems - @jdoliner?

Tryneus commented on src/server/pkg/transactionenv/env.go:
```diff
@@ -0,0 +1,87 @@
+// PpsTransactionServer is an interface for the transactionally-supported
+// methods that can be called through the PPS server.
+type PpsTransactionServer interface {
+	// TODO: add these once PPS is supported in transactions
```
This will be done in a future PR.

Tryneus commented on src/server/transaction/cmds/cmds.go:
```diff
@@ -0,0 +1,272 @@
+				return grpcutil.ScrubGRPC(err)
+			}
+			// TODO: use advisory locks on config so we don't have a race condition if
+			// two commands are run simultaneously
```
This will be done in a future PR - the worst that can happen from this is users mix up two transactions.

Tryneus commented on src/server/transaction/server/driver.go:
```diff
@@ -0,0 +1,292 @@
+		case *transaction.TransactionRequest_DeleteAll:
+			// TODO: extend this to delete everything through PFS, PPS, Auth and
+			// update the client DeleteAll call to use only this, then remove unused
+			// RPCs.
```
This will be done in a future PR after PPS support is added.

jdoliner commented on src/client/transaction/transaction.proto:
```diff
@@ -0,0 +1,85 @@
+}
+
+message TransactionRequest {
+  oneof request {
```
Up until this point we've opted to not use `oneof` and just do an "implicit oneof" by having a message with multiple fields in which only one should ever be set but that isn't enforced by the type checker. If I recall correctly we made that decision because the code required to interact with `oneof`s in go was pretty cumbersome and didn't seem worth the extra safety. We might want to rethink that but I think it'd be good for this code to be unified with what we already have and the path of least resistance would be to change this code rather than all the other stuff.

Tryneus replied:
I found the code for dealing with `oneof` to be not too cumbersome, at least vs trying to enforce such an invariant in code.  That is, when we add more supported requests and responses to the message, I would expect non-`oneof` code to need to scale O(n^2), while `oneof` code should scale O(n).  I could be wrong, though, if you have some prior art you can point me to.  I'm not opposed to changing this, but I don't want the code to become unmaintainable.

jdoliner commented on src/server/pfs/cmds/cmds.go:
```diff
@@ -71,13 +72,23 @@ or type (e.g. csv, binary, images, etc).`,
+				}
+			} else {
+				_, err = c.PfsAPIClient.CreateRepo(c.Ctx(), request)
+			}
```
Would it be possible to implement this server-side rather than clientside? We generally try to have as little logic as possible in the cmds, and it seems like this could be done pretty generically by checking the `ctx` to see if there's a txn in it or not server side.

Tryneus replied:
That sounds ok to me, worst that might happen is I end up with a circular dependency and need to shuffle some code around.  I'll give it a shot.

jdoliner commented on src/server/transaction/pretty/pretty.go:
```diff
@@ -0,0 +1,178 @@
+
+const (
+	//TransactionHeader is the header for transactions.
+	TransactionHeader = "TRANSACTION\tSTARTED\tLENGTH\t\n"
```
What would you think of renaming `LENGTH` to `OPS` or something like that? My immediate thought when I read "LENGTH" was that it was about the length of time the txn had been running for, (which it pretty obviously isn't, since that's what `STARTED` is), but I still think it might be a bit too generic of a name.

Tryneus replied:
Absolutely - I flip-flopped on this a bit before landing on length, but `OPS` sounds much better.

jdoliner commented on src/server/transaction/server/driver.go:
```diff
@@ -0,0 +1,292 @@
+	ctx := pachClient.Ctx()
+
+	// Run this thing in a loop in case we get a conflict, time out after some tries
+	for i := 0; i < 10; i++ {
```
It might be good the have the number of retries be configurable in the request. Just in case there's a situation where people have a lot of traffic and need this to be higher.

Tryneus replied:
Unfortunately because of [this change](https://github.com/pachyderm/pachyderm/pull/3658#discussion_r278336605), we won't be able to specify the number of retries on each request, it will need to be specified when creating or finishing the transaction.  I could add another `update transaction` command that could set this value, but it seems like overkill - thoughts?

jdoliner replied:
Yeah that seems like overkill, and configurable retries doesn't actually seem that important thinking about it more.

jdoliner commented on src/server/transaction/server/driver.go:
```diff
@@ -0,0 +1,292 @@
+					return &transactionConflictError{}
+				}
+
+				info.Requests = append(info.Requests, items...)
```
So am I right in understanding that if someone does a big put-file we'll put the request objects (which contain the content) into etcd? If so that's going to be a pretty bad bottleneck if people try to transactionally ingress large amounts of data. I don't think we should fix this as part of this PR but we should probably document it as a potential gotcha. And long term we'll need to figure out how to store that data elsewhere, which I don't think we'll be too hard.

Tryneus replied:
`put file` wasn't actually tackled in this PR for this reason - it needs another path from the supported operations.  The current plan is to upload the files to object storage, then store the references (whatever is needed to update the tree) into the transaction row in etcd.

jdoliner commented:
One thing I forgot to ask in the my first pass. Do batched calls to functions such as `StartCommit` which call `propagateCommit` batch their calls to `propagateCommit`? I.e. if I have a transaction that starts 2 commits in different branches and there's a pipeline that takes both of those branches as inputs will you get 1 output commit there or 2?

Tryneus commented:
This PR depends on #3718, but the tests in the integration branch (`grey/transaction-integration`) are passing, so I think this is ready for review @jdoliner.

In order to properly batch `propagateCommit` calls, there were a number of changes:
* `propagateCommit` -> `propagateCommits`
  * now takes an array of branches - this array and the subvenant branches are deduped before the actual work loop
* PFS no longer calls `propagateCommit` directly, but uses a new object `pfs.TransactionDefer`, which collects propagations and will run them at the end of a transaction
* `transactionenv` now provides its own wrapper around `col.NewSTM` that will construct a `transactionenv.TransactionContext` containing all the objects associated with the request/transaction:
  * client context
  * client for making inner api requests
  * `STM`
  * `pfs.TransactionDefer`
* Most writes in the `pfs.driver` layer were changed to accept a `TransactionContext`
* Cleaned up a few more places that were not making consistent requests in PFS during a transaction
* Dropped `pachctl copy file` and `pachctl delete file` support from transactions.
  * Due to the inability to use list operations in an STM, we can't invoke these on open commits
  * Technically we could allow them on closed commits, but considering these operations aren't the most valuable at the moment, it's better to avoid the confusion and technical debt for now.

jdoliner commented on src/client/transaction.go:
```diff
@@ -0,0 +1,298 @@
+	}
+	return nil, fmt.Errorf("added to transaction but received an unexpected response type")
+}
+*/
```
These should probably be removed or uncommented, it'll bitrot if it's left like this.

Tryneus replied:
:+1: - this was just old code I left in for reference while refactoring - deleted.

jdoliner commented on src/server/pkg/transactionenv/env.go:
```diff
@@ -0,0 +1,373 @@
+// the end of the transaction
+func (t *TransactionContext) PfsDefer() PfsTransactionDefer {
+	return t.pfsDefer
+}
```
Might as well make `pfsDefer` public if we're going to have an accessor method.

Tryneus replied:
Done (as well as the other getters in this struct)

jdoliner commented on src/server/pfs/server/transaction_defer.go:
```diff
@@ -0,0 +1,44 @@
+
+	// Branches to propagate when the transaction completes
+	propagateBranches []*pfs.Branch
+}
```
The name defer seems a little bit weird here because as far as I could tell this doesn't actually get deferred anywhere. I'm not sure I have a better name though.

Tryneus replied:
Renamed to `Propagater` so we can defer the bikeshedding until this type actually needs to do anything else but propagate commits.

jdoliner commented on src/server/transaction/server/driver.go:
```diff
@@ -0,0 +1,282 @@
+func now() *types.Timestamp {
+	t, err := types.TimestampProto(time.Now())
+	if err != nil {
+		panic(err)
```
Let's not panic here, it's unlikely to ever get hit but it makes me a bit nervous to think we could put the server in a crash loop with this, we can just swallow the error instead and return a zero time.

Tryneus replied:
Fixed here and in PFS where this was copied from.
